📝 Create slackctl package and write its documentation (#27)
IBM/slack-wrench: packages/slackctl/README.md

# slackctl

A utility for managing Slack [App](https://api.slack.com/start/overview) configurations as code with helpful integrations to make dev environments easy.

⚠️ Slack doesn't currently support this. To make this project work, we rely heavily on undocumented apis and client tokens. These tokens should be protected very carefully. Use at your own risk. This is not intended to be used for production, though it does make dev a whole lot easier.

# Table of Contents

- [Install](#install)
- [Updating a Slack App](#updating-a-slack-app)
- [App Configuration File](#app-configuration-file)
  - [Setting up a `local` config](#setting-up-a-local-config)
- [Getting a session token](#getting-a-session-token)

## Install

> Install the package globally for usage as a cli.

Reviewer: seems like maybe a good candidate for `npx`? then people could install locally easily also. I always dislike "requiring" global installs where avoidable.

```bash
yarn global add @slack-wrench/slackctl
# or
npm install -g @slack-wrench/slackctl
```

## Updating a Slack App

```bash
slackctl apply [flags]
```

Update an existing Slack app with a configuration. It will check the existing configuration, and apply it to match the configuration file. If you made any changes using the UI without updating the config, they will be overridden.

If your event subscription url needs changing, that url must be active and able to respond to a [`url_verification` event](https://api.slack.com/events/url_verification). If you're using the ngrok flag, then your local app should be running before trying to apply the config.

If you are using the ngrok flag, the command will wait for the specific localhost port to become active before trying to apply configuration.

You will need a [session token](#getting-a-session-token) to run this command. These tokens are not supported via Slack, use caution to keep them safe.

You can it via the environment variable `SLACK_SESSION_TOKEN`, a [local config](#setting-up-a-local-config), or via a flag (--session-token)

**Flags:**

> - `-s,--session-token=''` - A [session token](#getting-a-session-token) used for authentication

Author: Agreed, I'll remove it.

- `-n,--ngrok='3000'` - Start an [ngrok tunnel](https://ngrok.com/) to your localhost. This will also modify your configuration to change urls to the url exposed from ngrok.
- `-t,--timeout='10'` - Used with the `--ngrok` flag, how long to wait for your localhost app to become active before timing out.

```bash
slackctl validate
```

> 

Reviewer: other potentially relevant commands could be `dry-run` and/or `diff` - which would check/report on differences between local and deployed.

Author: I do like those, Is a good question re: how to merge this as it won't be 100% accurate for a while me thinks.

Reviewer: see #33

This will read your configuration and validate that it meets the correct schema.

## App Configuration File

> Use this file to define your app's oauth scopes, commands, and app urls you would normally do on your App's [configuration page](https://api.slack.com/apps).

Reviewer: ```suggestion
Use this file to define your app's oauth scopes, commands, and app urls you would normally manually set in your App's [configuration page](https://api.slack.com/apps).
```

> This file should be in the root of your project and be named one of `slack.config.yaml`, `slack.config.yml`, `slack.config.js`, or `slack.config.json`.

Author: Because it's easy with `cosmiconfig` and I wanted `.js` for me, but see adoption in python land projects scary if not a `.yaml`

> Example:

Author: Probably, I want to start with the max config and filter it down as we learn. I'm honestly not sure what that is right now.

# you don't need to include the section
interactiveComponents:
  # Any interactions with actions, modals, or interactive components (such as buttons, select menus, and datepickers) will be sent to a URL you specify.
  requestPath: /slack/events

  # Edit message actions and their callbacks
  actions:
    # A good name is short and descriptive. Please use sentence case (e.g. File a bug, not File A Bug).
    - name: File a Bug
      # A short description for your action
      description: Creates a bug report in Super App
      # Your app uses this to identify the action. It’s not visible to users.
      callbackId: action_name

  # If you want to load external data in your select menus, you need to provide an options load URL.
  optionsLoadPath: /slack/events

slashCommands:
  - command: /command
    requestPath: /slack/events
    description: Launches the Rocket!
    # Optionally list any parameters that can be passed.
    hint: '[which rocket to launch]'
    # Escape channels, users, and links sent to your app
    escape: true

oauth:
  # OAuth scopes let you specify exactly how your app needs to access a Slack user’s account.

>   scopes:

Author: With granular scopes rolling out of beta before I expected, I'll rework this.

  requestPath: /slack/events

  # Bot users can subscribe to events related to the channels and conversations they're part of.
  botEvents:
    - app_mention
    - message.channels
    - message.groups

  # To subscribe to an event, your app must have access to the related OAuth permission scope.
  workspaceEvents:
    - message.im

  # Your app can be notified when users post links from up to 5 domains. This requires the links:read scope and a subscription to the link_shared event.
  # You'll receive events for any subdomain, or path of the selected domains or hosts. Ports are ignored, so requests on any port will trigger an event. Using IP addresses is not allowed.
  unfurlDomains:
    - ibm.com
    - twitter.com

bot:
  # Names must be shorter than 80 characters, and can’t use punctuation (other than apostrophes and periods).
  displayName: Botty McBotter
  # If this username isn’t available on any workspace that tries to install it, we will slightly change it to make it work. Usernames must be all lowercase. They cannot be longer than 21 characters and can only contain letters, numbers, periods, hyphens, and underscores.
  defaultUsername: botty

>   # When this is false, Slack automatically displays whether your bot is online based on usage of the RTM API.

Author: 😛 Yeah I copied all text from Slack's UI. Probably worth clarifying.

>   # When this is false, Slack automatically displays whether your bot is online based on usage of the RTM API.

Author: With the new rollout of granular scopes they changed the UI.. 

Now under the `App Home` tab and they removed all descriptions 😛
